dali/test/op_graph.h

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_TEST_OP_GRAPH_H_
+#define DALI_TEST_OP_GRAPH_H_
+
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+#include <iostream>
+
+namespace dali {
+namespace testing {
+
+struct Node;
+
+struct Edge {
+  Edge() = default;
+  Edge(Node *n, int index, bool explicit_GPU = false) : node(n), index(index), explicit_GPU(false) {}
+
+  Node *node = nullptr;
+  int index = 0;
+  bool explicit_GPU = false;
+
+  Edge gpu() const {
+    auto ret = *this;
+    ret.explicit_GPU = true;
+    return ret;
+  }
+};
+
+struct Node {
+  Node() = default;
+  Node(const Node &) = delete;
+  Node(Node &&) = default;
+
+  Node(std::string op_name, std::string node_name, int backend_mask)
+  : op_name(std::move(op_name)), node_name(std::move(node_name)), backend_mask(backend_mask) {}
+
+  template <typename... Inputs>
+  Node &operator()(Inputs&&... inputs) {
+    int junk[] = { (add_in(inputs), 0)... };
+    (void)junk;
+    return *this;
+  }
+
+  void add_in(Edge e) {
+    in.push_back(e);
```

Comment:
Don't we need to move this `in.push_back(e);` to the end after we modify e, not before?

Reply:
Oh, I see, we modify only things pointed by node.